pootle/core/models/snapshot.py

```diff
@@ -0,0 +1,159 @@
+
+    @cached_property
+    def reference(self):
+        try:
+            with open(self.filepath) as reference_file:
+                return reference_file.read()
+        except IOError:
+            return None
+
+    def clean(self, data):
+        """Cleans up `data` before using it as a snapshot reference."""
+        # XXX: maybe we can do something smarter than blacklisting when we
+        # have a `ContextList`?
+        if isinstance(data, dict) or isinstance(data, ContextList):
+            return {
+                key: self.clean(data[key]) for key in data.keys()
+                if key not in BLACKLISTED_KEYS
+            }
+        elif isinstance(data, list):
+            return [self.clean(item) for item in data]
+        return data
+
+    def serialize(self, ctx_data):
+        return jsonify(self.clean(ctx_data), indent=2)
+
+    def save(self, ctx_data):
+        """Saves snapshot reference file to disk."""
+        # Make sure directory exists before saving to file
+        try:
+            os.makedirs(os.path.dirname(self.filepath))
+        except OSError:
+            pass
+
+        with open(self.filepath, 'w') as outfile:
+            outfile.write(ctx_data)
+
+    def assert_matches(self, ctx_data):
+        """Asserts whether `ctx_data` matches the reference snapshot.
+
+        The `ctx` data will be cleaned prior to being stored or compared
+        against any references.
+
+        Implementation detail: the reason why this function asserts and
+        doesn't return a boolean is to have useful diffs upon assertion
+        failures.
+        It would be possible to implement custom pytest assertion classes
+        but that's beyond the simplest implementation :)
+        """
+        ctx_data = self.serialize(ctx_data)
+
+        if self.generate:
+            self.save(ctx_data)
+            return
+
+        if self.reference is None:
+            raise AssertionError(
+                'Snapshot file "%s" does not exist.' % self.filepath
+            )
+
+        assert self.reference == ctx_data
+
+
+class SnapshotStack(object):
+    """Context processor implementing a simple stack of `Snapshot`s.
+
+    >> s = SnapshotStack('/foo/bar/')
+    >> with s.push('foo'):
+    >>    # context is 'foo'
+    >>    with s.push('bar'):
+    >>        # context is 'foo.bar'
+
+    Pushing a list of values will be flattened to individual values when
+    determining the context:
+
+    >> with s.push(['foo', 'bar']):
+    >>    # context is 'foo.bar'
+    """
+
+    def __init__(self, snapshot_kwargs):
+        self.stack = []
+        self.snapshot_kwargs = snapshot_kwargs
+
+    def __enter__(self):
+        return Snapshot(self.ctx_name, **self.snapshot_kwargs)
+
+    def __exit__(self, *args, **kwargs):
+        self.pop()
+
+    @property
+    def ctx_name(self):
+        # Disallow forward slashes
+        stack_items = [
+            item.replace('/', '', 1) if item.startswith('/') else item
```

Comment:
We should be also safeguarding ourselves from paths containing relative paths like `../../`. In no circumstance the code should create directories and files outside its root snapshot directory. Probably it makes sense to fail the test if `..` is found in the file path.